### PR TITLE
Use WC built-in Action Scheduler to fetch in-app promotions

### PR DESCRIPTION
**Pull request by corsonr**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

In https://github.com/woocommerce/woocommerce/pull/44655 and related PRs, we introduced logic to fetch data from a new Woo.com endpoint and show notices on the WooCommerce Marketplace pages, as well as a bubble on the `WooCommerce > Extensions` menu item. That logic, based on events scheduled with `wp_schedule_event`, didn't run on JN and other external sites, so the notices weren't shown. In this PR, we are using ActionScheduler to schedule the event with `as_schedule_recurring_action`.

Closes 19763-gh-Automattic/woocommerce.com.

#### If you're using a local `wp-env` environment

1. Check out this branch locally.
2. Edit `plugins/woocommerce/includes/admin/class-wc-admin-marketplace-promotions.php` and replace the `PROMOTIONS_API_URL` value `https://woo.com/wp-json/wccom-extensions/3.0/promotions` with `https://gist.githubusercontent.com/andfinally/acd8646151c0108f92979c985c20a0d8/raw/`. This is a gist with dummy data that is valid until 31 March.
3. On line 21, change the frequency of the scheduled action to five minutes:

```php
const SCHEDULED_ACTION_INTERVAL = 5 * MINUTE_IN_SECONDS;
```
4. Run `nvm use`.
5. Do `cd plugins/woocommerce`.
6. Run `pnpm -- wp-env start` to start wp-env.
7. run `pnpm --filter='@woocommerce/plugin-woocommerce' build`.
8. Visit the [pending scheduled actions page](http://localhost:8888/wp-admin/admin.php?page=wc-status&tab=action-scheduler&status=pending) and search for `woocommerce_marketplace_fetch_promotions`. Note that the action is pending.
9. After 5 minutes or so, the action should run. (You may need to jog it by viewing a page on your test site.)
10. The menu bubble should appear on the `WooCommerce > Extensions` menu item, and the notice should appear on the default Marketplace page `http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fextensions`. (It also appears on the Themes and other tabs on that page – that's intentional.)

<p>
image
</p>

11. Edit `class-wc-admin-marketplace-promotions.php` again and change the `PROMOTIONS_API_URL` to `https://gist.githubusercontent.com/andfinally/8ffca3fe3844fee6e146a80e563f00bb/raw/`. This is a second gist, where the expiry dates for the promotions have passed.
12. Wait for the scheduled action to run again.
13. The notice and menu bubble should not be shown.

#### If you're using a JN or other external site

1. Do the steps 1-5 as for the local environment.
2. Comment out this line in `plugins/woocommerce/bin/build-zip.sh`. This skips a memory-intensive process in the WooCommerce zip build that is irrelevant to this change.

```
pnpm --filter=@woocommerce/plugin-woocommerce makepot || exit "$?"
```

3. Run the zip build: `pnpm --filter='@woocommerce/plugin-woocommerce' build:zip`.
4. When the build is complete, you should see the resulting `woocommerce.zip` plugin file in `plugins/woocommerce`. Install the plugin in your external site.
5. The menu bubble should appear on the `WooCommerce > Extensions` menu item, and the notice should appear on the `wp-admin/admin.php?page=wc-admin&path=%2Fextensions` page.
6. Change the `PROMOTIONS_API_URL` to `https://gist.githubusercontent.com/andfinally/8ffca3fe3844fee6e146a80e563f00bb/raw/`. This is a second gist, where the expiry dates for the promotions have passed. If you have SSH access to your site, you can make this change directly in the PHP file.
7. Otherwise, build the zip again as in step 3, deactivate and delete the WooCommerce plugin on your site, and install the new zip.
8. Wait for the scheduled action to run. The notice and menu item badge should no longer appear.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/includes/admin/class-wc-admin-marketplace-promotions.php [hunk 1]
```diff
@@ -26,69 +27,68 @@ class WC_Admin_Marketplace_Promotions {
...
-			if ( false === wp_next_scheduled( self::SCHEDULED_ACTION_HOOK ) ) {
-				wp_schedule_event( time(), 'twicedaily', self::SCHEDULED_ACTION_HOOK );
-			}
+		add_action( 'init', array( __CLASS__, 'schedule_promotion_fetch' ), 10 );
 
-			self::$locale = ( self::$locale ?? get_user_locale() ) ?? 'en_US';
-			self::maybe_show_bubble_promotions();
+		if (
+			defined( 'DOING_AJAX' ) && DOING_AJAX
+			|| defined( 'DOING_CRON' ) && DOING_CRON
```
Comment (mcliwanow) [inline plugins/woocommerce/includes/admin/class-wc-admin-marketplace-promotions.php:65]: should we also check for CLI ?
```php
defined( 'WP_CLI' ) && WP_CLI
```

**Review by mcliwanow — APPROVED** [review]
Everything works as expected. Action is triggered when needed, the specified endpoint is called and both bubble and notice are visible.

File: plugins/woocommerce/includes/admin/class-wc-admin-marketplace-promotions.php (hunk 1 above)
Comment (corsonr) [reply plugins/woocommerce/includes/admin/class-wc-admin-marketplace-promotions.php:65]: Done in https://github.com/woocommerce/woocommerce/pull/45628/commits/4705f893a0aa60987126f5712a3258b03f1ca1c3

**Comment by andfinally** [comment]
Thanks for the review Michal! 🙌 

I've committed a small change to ensure the scheduled action works when run by cron. 

* We were including and instantiating `WC_Admin_Marketplace_Promotions` in the `WC_Admin` constructor. 
* But `WC_Admin` is only instantiated during `is_admin` requests. We also need to respond to cron requests. 
* So we're now including the class from `class-woocommerce.php` if the context is admin or cron, and instantiating it on `init`. 
* This fixes this error in the Scheduled Actions:

`action failed via WP Cron: Scheduled action for woocommerce_marketplace_fetch_promotions will not be executed as no callbacks are registered.`

**Review by andfinally — APPROVED** [review]
Looks good! 👍 Tested on a new JN site and things worked as expected. Also tested that a `woocommerce_marketplace_suppress_promotions` filter cancels the scheduled action.

**Pull request merged**